src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,271 @@
+  typedef struct mmm_mem_stat
+  {
+    std::atomic<uint64_t> init_stat;
+    std::atomic<uint64_t> cur_stat;
+    std::atomic<uint64_t> peak_stat;
+    std::atomic<uint32_t> expand_count;
+  } MMM_MEM_STAT;
+
+  MMM_COMP_INFO heap_comp_info[] =
+  {
+    {HEAP, "", ""},
+    {HEAP_ATTRINFO, "ATTRINFO CACHE", ""}
+  };
+
+  class mmm_subcomponent
+  {
+    public:
+      mmm_subcomponent (const char *name)
+      {
+	m_subcomp_name = new char[strlen (name) + 1];
+	strcpy (m_subcomp_name, name);
+      }
+
+      /* public variable */
+      std::atomic<uint64_t> m_cur_stat;
+
+      /* function */
+      char *get_name ()
+      {
+	return m_subcomp_name;
+      }
+
+      ~mmm_subcomponent ()
+      {
+	delete[] m_subcomp_name;
+      }
+    private:
+      /* private variable */
+      char *m_subcomp_name;
+  };
+
+  class mmm_component
+  {
+    public:
+      mmm_component (const char *name)
+      {
+	m_comp_name = new char[strlen (name) + 1];
+	strcpy (m_comp_name, name);
+      }
+
+      /* public variable */
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_subcomponent> m_subcomponent;
+
+      /* function */
+      int add_subcomponent (const char *name);
+      char *get_name ()
+      {
+	return m_comp_name;
+      }
+
+      ~mmm_component ()
+      {
+	delete[] m_comp_name;
+      }
+    private:
+      /* private variable */
+      char *m_comp_name;
+  };
+
+  class mmm_module
+  {
+    public:
+      mmm_module (const char *name, MMM_COMP_INFO *info);
+      mmm_module () {} /* for dummy */
+
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int m_max_idx = 0x0000FFFF;
+
+      /* public variable */
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_component> m_component;
+      std::vector<int> m_comp_idx_map;
+
+      /* function */
+      virtual int aggregate_stats (MEMMON_MODULE_INFO *info);
+      int add_component (const char *name);
+      char *get_name ()
+      {
+	return m_module_name;
+      }
+      static inline int MMM_MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx)
+      {
+	return (comp_idx << 16 | subcomp_idx);
+      }
+
+      ~mmm_module ()
+      {
+	if (m_module_name)
+	  {
+	    delete[] m_module_name;
+	  }
+      }
+    private:
+      /* private variable */
+      char *m_module_name;
+  };
+
+  class mmm_aggregater
+  {
+    public:
+      mmm_aggregater (memory_monitoring_manager *mmm)
+      {
+	this->m_memmon_mgr = mmm;
+      }
+
+      /* function */
+      int get_server_info (MEMMON_SERVER_INFO *info);
+      int get_module_info (MEMMON_MODULE_INFO *info, int module_index);
+      int get_transaction_info (MEMMON_TRAN_INFO *info, int tran_count);
+
+      ~mmm_aggregater () {}
+    private:
+      /* private variable */
+      /* backlink of memory_monitoring_manager class */
+      memory_monitoring_manager *m_memmon_mgr;
```

Comment:
```suggestion
      memory_monitoring_manager &m_memmon_mgr;
```
Ref is better than pointer in this case. No danger of wrong access and dangling. I think we can even define `const memory_monitoring_manager *m_memmon_mgr;` since the aggregate doesn't modify it, I guess.